// react-amap/lib/polyline/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _isFun = require('../utils/isFun');

var _isFun2 = _interopRequireDefault(_isFun);

var _log = require('../utils/log');

var _log2 = _interopRequireDefault(_log);

var _polyeditor = require('../polyeditor');

var _polyeditor2 = _interopRequireDefault(_polyeditor);

var _toCapitalString = require('../utils/toCapitalString');

var _toCapitalString2 = _interopRequireDefault(_toCapitalString);

var _common = require('../utils/common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Component = React.Component;

var configurableProps = ['path', 'extData', 'draggable',

/* 扩展属性*/
'visible', 'style'];

var allProps = configurableProps.concat(['zIndex', 'bubble', 'showDir']);

var Polyline = function (_Component) {
  _inherits(Polyline, _Component);

  function Polyline(props) {
    _classCallCheck(this, Polyline);

    var _this = _possibleConstructorReturn(this, (Polyline.__proto__ || Object.getPrototypeOf(Polyline)).call(this, props));

    if (typeof window !== 'undefined') {
      if (!props.__map__) {
        _log2.default.warning('MAP_INSTANCE_REQUIRED');
      } else {
        _this.map = props.__map__;
        _this.element = _this.map.getContainer();
        _this.createMapPolyline(props);
      }
    }
    return _this;
  }

  _createClass(Polyline, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.map) {
        this.refreshPolylineLayout(nextProps);
      }
    }
  }, {
    key: 'createMapPolyline',
    value: function createMapPolyline(props) {
      var options = this.buildCreateOptions(props);
      options.map = this.map;
      this.polyline = new window.AMap.Polyline(options);

      var events = this.exposeLineInstance(props);
      events && this.bingLineEvents(events);

      if ('visible' in props) {
        if (props.visible) {
          this.polyline.show();
        } else {
          this.polyline.hide();
        }
      }
    }
  }, {
    key: 'buildCreateOptions',
    value: function buildCreateOptions(props) {
      var _this2 = this;

      var options = {};
      allProps.forEach(function (key) {
        if (key in props) {
          if (key === 'style' && props.style) {
            var styleItem = Object.keys(props.style);
            styleItem.forEach(function (item) {
              // $FlowFixMe
              options[item] = props.style[item];
            });
            // visible 做特殊处理
          } else if (key !== 'visible') {
            options[key] = _this2.getSetterValue(key, props[key]);
          }
        }
      });
      return options;
    }
  }, {
    key: 'refreshPolylineLayout',
    value: function refreshPolylineLayout(nextProps) {
      var _this3 = this;

      configurableProps.forEach(function (key) {
        if (key in nextProps) {
          if (_this3.detectPropChanged(key, nextProps)) {
            if (key === 'visible') {
              if (nextProps.visible) {
                _this3.polyline.show();
              } else {
                _this3.polyline.hide();
              }
            } else if (key === 'style') {
              _this3.polyline.setOptions(nextProps.style);
            } else {
              var setterName = 'set' + (0, _toCapitalString2.default)(key);
              var setterValue = _this3.getSetterValue(key, nextProps[key]);
              _this3.polyline[setterName](setterValue);
            }
          }
        }
      });
    }
  }, {
    key: 'detectPropChanged',
    value: function detectPropChanged(key, nextProps) {
      return this.props[key] !== nextProps[key];
    }
  }, {
    key: 'getSetterValue',
    value: function getSetterValue(key, value) {
      if (key === 'path') {
        return this.buildPathValue(value);
      }
      return value;
    }
  }, {
    key: 'buildPathValue',
    value: function buildPathValue(path) {
      if (path.length) {
        if ('getLng' in path[0]) {
          return path;
        }
        return path.map(function (p) {
          return (0, _common.getAMapPosition)(p);
        });
      }
      return path;
    }
  }, {
    key: 'exposeLineInstance',
    value: function exposeLineInstance(props) {
      if ('events' in props) {
        var _events = props.events;
        if ((0, _isFun2.default)(_events.created)) {
          _events.created(this.polyline);
        }
        delete _events.created;
        return _events;
      }
      return false;
    }
  }, {
    key: 'bingLineEvents',
    value: function bingLineEvents(events) {
      var _this4 = this;

      var list = Object.keys(events);
      list.length && list.forEach(function (evName) {
        _this4.polyline.on(evName, events[evName]);
      });
    }
  }, {
    key: 'renderEditor',
    value: function renderEditor(children) {
      if (!children) {
        return null;
      }
      if (React.Children.count(children) !== 1) {
        return null;
      }
      var child = React.Children.only(children);
      if (child.type === _polyeditor2.default) {
        return React.cloneElement(child, {
          __poly__: this.polyline,
          __map__: this.map,
          __ele__: this.element
        });
      }
      return null;
    }
  }, {
    key: 'render',
    value: function render() {
      return this.renderEditor(this.props.children);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.polyline.hide();
      this.polyline.setMap(null);
      delete this.polyline;
    }
  }]);

  return Polyline;
}(Component);

exports.default = Polyline;
