// react-amap/lib/circle/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _isFun = require('../utils/isFun');

var _isFun2 = _interopRequireDefault(_isFun);

var _log = require('../utils/log');

var _log2 = _interopRequireDefault(_log);

var _toCapitalString = require('../utils/toCapitalString');

var _toCapitalString2 = _interopRequireDefault(_toCapitalString);

var _common = require('../utils/common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

/*
 * props
 * {
 *  __map__ 父级组件传过来的地图实例
 * }
 */

var configurableProps = ['center', 'radius', 'draggable', 'extData',

/* 原生的接口中并没有这些对象，这是本组件的扩展 */
'visible', 'style'];

var allProps = configurableProps.concat(['zIndex', 'bubble']);

var Circle = function (_React$Component) {
  _inherits(Circle, _React$Component);

  function Circle(props) {
    _classCallCheck(this, Circle);

    var _this = _possibleConstructorReturn(this, (Circle.__proto__ || Object.getPrototypeOf(Circle)).call(this, props));

    if (typeof window !== 'undefined') {
      if (!props.__map__) {
        _log2.default.warning('MAP_INSTANCE_REQUIRED');
      } else {
        _this.map = props.__map__;
        _this.element = _this.map.getContainer();
        _this.initMapCircle(props);
      }
    }
    return _this;
  }

  _createClass(Circle, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.map) {
        this.refreshCircleLayout(nextProps);
      }
    }
  }, {
    key: 'initMapCircle',
    value: function initMapCircle(props) {
      var options = this.buildCreateOptions(props);
      options.map = this.map;
      this.mapCircle = new window.AMap.Circle(options);
      var events = this.exposeCircleInstance(props);
      events && this.bindCircleEvents(events);

      if ('visible' in props) {
        if (!props.visible) {
          this.mapCircle.hide();
        }
      }
    }
  }, {
    key: 'buildCreateOptions',
    value: function buildCreateOptions(props) {
      var _this2 = this;

      var options = {};
      allProps.forEach(function (key) {
        if (key in props) {
          if (key === 'style' && props.style !== undefined) {
            var styleItem = Object.keys(props.style);
            styleItem.forEach(function (item) {
              // $FlowFixMe
              options[item] = props.style[item];
            });
          } else {
            if (key !== 'visible') {
              options[key] = _this2.getSetterValue(key, props);
            }
          }
        }
      });
      return options;
    }
  }, {
    key: 'refreshCircleLayout',
    value: function refreshCircleLayout(nextProps) {
      var _this3 = this;

      configurableProps.forEach(function (key) {
        if (key in nextProps) {
          if (_this3.checkPropChanged(key, nextProps)) {
            if (key === 'visible') {
              if (nextProps.visible) {
                _this3.mapCircle.show();
              } else {
                _this3.mapCircle.hide();
              }
            } else if (key === 'style') {
              _this3.mapCircle.setOptions(nextProps.style);
            } else {
              var setterName = 'set' + (0, _toCapitalString2.default)(key);
              var setterParam = _this3.getSetterValue(key, nextProps);
              _this3.mapCircle[setterName](setterParam);
            }
          }
        }
      });
    }
  }, {
    key: 'checkPropChanged',
    value: function checkPropChanged(key, nextProps) {
      return this.props[key] !== nextProps[key];
    }
  }, {
    key: 'getSetterValue',
    value: function getSetterValue(key, props) {
      if (key === 'center') {
        return (0, _common.getAMapPosition)(props.center);
      }
      return props[key];
    }
  }, {
    key: 'exposeCircleInstance',
    value: function exposeCircleInstance(props) {
      if ('events' in props) {
        var _events = props.events || {};
        if ((0, _isFun2.default)(_events.created)) {
          _events.created(this.mapCircle);
        }
        delete _events.created;
        return _events;
      }
      return false;
    }
  }, {
    key: 'bindCircleEvents',
    value: function bindCircleEvents(events) {
      var _this4 = this;

      var list = Object.keys(events);
      list.length && list.forEach(function (evName) {
        _this4.mapCircle.on(evName, events[evName]);
      });
    }
  }, {
    key: 'renderEditor',
    value: function renderEditor(children) {
      if (!children) {
        return null;
      }
      if (React.Children.count(children) !== 1) {
        return null;
      }
      return React.cloneElement(React.Children.only(children), {
        __circle__: this.mapCircle,
        __map__: this.map,
        __ele__: this.element
      });
    }
  }, {
    key: 'render',
    value: function render() {
      return this.renderEditor(this.props.children);
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.mapCircle.hide();
      this.mapCircle.setMap(null);
      delete this.mapCircle;
    }
  }]);

  return Circle;
}(React.Component);

exports.default = Circle;
