// react-amap/lib/map/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _APILoader = require('../utils/APILoader');

var _APILoader2 = _interopRequireDefault(_APILoader);

var _isFun = require('../utils/isFun');

var _isFun2 = _interopRequireDefault(_isFun);

var _log = require('../utils/log');

var _log2 = _interopRequireDefault(_log);

var _toCapitalString = require('../utils/toCapitalString');

var _toCapitalString2 = _interopRequireDefault(_toCapitalString);

var _common = require('../utils/common');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Component = React.Component;
var Children = React.Children;
var containerStyle = {
  width: '100%',
  height: '100%'
};
var wrapperStyle = {
  width: '100%',
  height: '100%',
  position: 'relative'
};

// Native supported dynamic props by Amap
var NativeDynamicProps = ['layers', 'zoom', 'center', 'labelzIndex',

// 'lang', native error in JSSDK when 3D viewMode
'mapStyle', 'features', 'cursor', 'pitch'];

var ExtendedDynamicProps = ['city', 'bounds', 'limitBounds', 'status', 'rotation'];

/*
 * Props below can set by 'setStatus' altogether
 */
var StatusDynamicProps = ['animateEnable', 'doubleClickZoom', 'dragEnable', 'isHotspot', 'jogEnable', 'keyboardEnable', 'resizeEnable', 'rotateEnable', 'scrollWheel', 'touchZoom', 'zoomEnable'];

var StaticProps = ['view', 'zooms', 'showIndoorMap', 'indoorMap', 'expandZoomRange', 'showBuildingBlock', 'viewMode', 'pitchEnable', 'buildingAnimation', 'skyColor'];

var CreateProps = NativeDynamicProps.concat(StatusDynamicProps, StaticProps);

var defaultOpts = {
  MapType: {
    showRoad: false,
    showTraffic: false,
    defaultType: 0
  },
  ToolBar: {
    position: 'RB',
    noIpLocate: true,
    locate: true,
    liteStyle: true,
    autoPosition: false
  },
  OverView: {},
  ControlBar: {}
};

var Map = function (_Component) {
  _inherits(Map, _Component);

  function Map(props) {
    _classCallCheck(this, Map);

    var _this = _possibleConstructorReturn(this, (Map.__proto__ || Object.getPrototypeOf(Map)).call(this, props));

    _this.state = {
      mapLoaded: false
    };
    if (typeof window !== 'undefined') {
      _this.pluginMap = {};
      _this.loader = new _APILoader2.default(props.amapkey, props.useAMapUI).load();
    }
    return _this;
  }

  _createClass(Map, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      var _this2 = this;

      var prevProps = this.props;
      this.loader.then(function () {
        if (_this2.map) {
          _this2.updateMapProps(prevProps, nextProps);
        }
      });
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.loadMap();
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      this.loadMap();
    }
  }, {
    key: 'loadMap',
    value: function loadMap() {
      var _this3 = this;

      this.loader.then(function () {
        _this3.initMapInstance();
        if (!_this3.state.mapLoaded) {
          _this3.setState({
            mapLoaded: true
          });
        }
      });
    }
  }, {
    key: 'renderChildren',
    value: function renderChildren() {
      var _this4 = this;

      return Children.map(this.props.children, function (child) {
        if (child) {
          var cType = child.type;
          /* 针对下面两种组件不注入地图相关属性
           * 1. 明确声明不需要注入的
           * 2. DOM 元素
           */
          if (cType.preventAmap || typeof cType === 'string') {
            return child;
          }
          return React.cloneElement(child, {
            __map__: _this4.map,
            // consider to remove __ele__, because map.getContainer can also get this
            __ele__: _this4.mapWrapper
          });
        }
        return child;
      });
    }
  }, {
    key: 'initMapInstance',
    value: function initMapInstance() {
      var _this5 = this;

      if (!this.map) {
        var options = this.buildCreateOptions();
        this.map = new window.AMap.Map(this.mapWrapper, options);
        // event binding
        var events = this.exposeMapInstance();
        events && this.bindAMapEvents(events);
        // install map plugins
        this.setPlugins(this.props);
        // binding extended props
        ExtendedDynamicProps.forEach(function (key) {
          if (key in _this5.props) {
            var setterParam = _this5.getSetterValue(key, _this5.props);
            _this5.runMapSetter(key, setterParam);
          }
        });
      }
    }
  }, {
    key: 'buildCreateOptions',
    value: function buildCreateOptions() {
      var _this6 = this;

      var props = this.props;
      var options = {};
      CreateProps.forEach(function (key) {
        if (key in props) {
          options[key] = _this6.getSetterValue(key, props);
        }
      });
      return options;
    }
  }, {
    key: 'bindAMapEvents',
    value: function bindAMapEvents(events) {
      var _this7 = this;

      var list = Object.keys(events);
      list.length && list.forEach(function (evName) {
        _this7.map.on(evName, events[evName]);
      });
    }
  }, {
    key: 'updateMapProps',
    value: function updateMapProps(prevProps, nextProps) {
      var _this8 = this;

      var nextMapStatus = {};
      var statusChangeFlag = false;
      var statusPropExist = false;
      StatusDynamicProps.forEach(function (key) {
        if (key in nextProps) {
          statusPropExist = true;
          if (_this8.detectPropChanged(key, prevProps, nextProps)) {
            statusChangeFlag = true;
            nextMapStatus[key] = nextProps[key];
          }
        }
      });
      statusChangeFlag && this.map.setStatus(nextMapStatus);
      if (statusPropExist && 'status' in nextProps) {
        _log2.default.warning('\u4EE5\u4E0B\u8FD9\u4E9B\u5C5E\u6027\u53EF\u4EE5\u5355\u72EC\u63D0\u4F9B\u8FDB\u884C\u914D\u7F6E\uFF0C\u4E5F\u53EF\u4EE5\u7EDF\u4E00\u4F5C\u4E3A\u2018status\u2019\u5C5E\u6027\u914D\u7F6E\uFF1B\u4F46\u662F\u8BF7\u4E0D\u8981\u540C\u65F6\u4F7F\u7528\u8FD9\u4E24\u79CD\u65B9\u5F0F\u3002\n\uFF08' + StatusDynamicProps.join(', ') + '\uFF09');
      }
      NativeDynamicProps.concat(ExtendedDynamicProps).forEach(function (key) {
        if (key in nextProps) {
          if (_this8.detectPropChanged(key, prevProps, nextProps)) {
            var setterParam = _this8.getSetterValue(key, nextProps);
            _this8.runMapSetter(key, setterParam);
          }
        }
      });
      StaticProps.forEach(function (key) {
        if (key in nextProps) {
          if (_this8.detectPropChanged(key, prevProps, nextProps)) {
            _log2.default.warning('\'' + key + '\' \u662F\u4E00\u4E2A\u9759\u6001\u5C5E\u6027\uFF0C\u5730\u56FE\u5B9E\u4F8B\u521B\u5EFA\u6210\u529F\u540E\u65E0\u6CD5\u4FEE\u6539');
          }
        }
      });
      this.setPlugins(nextProps);
    }
  }, {
    key: 'runMapSetter',
    value: function runMapSetter(key, setterParam) {
      if (key === 'limitBounds' && !setterParam) {
        this.map.clearLimitBounds();
      } else {
        var setterName = this.getSetterName(key);
        this.map[setterName](setterParam);
      }
    }
  }, {
    key: 'getSetterValue',
    value: function getSetterValue(key, props) {
      if (key === 'center') {
        return (0, _common.getAMapPosition)(props.center);
      }
      return props[key];
    }
  }, {
    key: 'getSetterName',
    value: function getSetterName(key) {
      switch (key) {
        case 'labelzIndex':
          return 'setlabelzIndex';
        case 'cursor':
          return 'setDefaultCursor';
        default:
          return 'set' + (0, _toCapitalString2.default)(key);
      }
    }
  }, {
    key: 'detectPropChanged',
    value: function detectPropChanged(key, prevProps, nextProps) {
      return prevProps[key] !== nextProps[key];
    }
  }, {
    key: 'setPlugins',
    value: function setPlugins(props) {
      var _this9 = this;

      var pluginList = ['Scale', 'ToolBar', 'MapType', 'OverView', 'ControlBar'];
      if ('plugins' in props) {
        var plugins = props.plugins;
        if (plugins && plugins.length) {
          plugins.forEach(function (p) {
            var name = void 0,
                config = void 0,
                visible = void 0;
            if (typeof p === 'string') {
              name = p;
              config = null;
              visible = true;
            } else {
              name = p.name;
              config = p.options || {};
              visible = 'visible' in config && typeof config.visible === 'boolean' ? config.visible : true;
              delete config.visible;
            }
            var idx = pluginList.indexOf(name);
            if (idx === -1) {
              _log2.default.warning('\u6CA1\u6709 \u2018' + name + '\u2019 \u8FD9\u4E2A\u63D2\u4EF6\uFF0C\u8BF7\u68C0\u67E5\u662F\u5426\u62FC\u5199\u9519\u8BEF');
            } else {
              if (visible) {
                pluginList.splice(idx, 1);
                _this9.installPlugin(name, config);
              }
            }
          });
        }
      }
      this.removeOrDisablePlugins(pluginList);
    }
  }, {
    key: 'removeOrDisablePlugins',
    value: function removeOrDisablePlugins(plugins) {
      var _this10 = this;

      if (plugins && plugins.length) {
        plugins.forEach(function (p) {
          if (p in _this10.pluginMap) {
            // ControlBar has no 'hide' method
            if (p === 'ControlBar') {
              _this10.map.removeControl(_this10.pluginMap[p]);
              delete _this10.pluginMap[p];
            } else {
              _this10.pluginMap[p].hide();
            }
          }
        });
      }
    }
  }, {
    key: 'installPlugin',
    value: function installPlugin(name, opts) {
      opts = opts || {};
      switch (name) {
        case 'Scale':
        case 'ToolBar':
        case 'OverView':
        case 'MapType':
          this.setMapPlugin(name, opts);
          break;
        case 'ControlBar':
          this.setControlBar(opts);
          break;
        default:
        // do nothing
      }
    }
  }, {
    key: 'setMapPlugin',
    value: function setMapPlugin(name, opts) {
      var _this11 = this;

      if (this.pluginMap[name]) {
        this.pluginMap[name].show();
      } else {
        var onCreated = opts.onCreated,
            restOpts = _objectWithoutProperties(opts, ['onCreated']);

        var initOpts = _extends({}, defaultOpts[name], restOpts);
        this.map.plugin(['AMap.' + name], function () {
          _this11.pluginMap[name] = new window.AMap[name](initOpts);
          _this11.map.addControl(_this11.pluginMap[name]);
          if ((0, _isFun2.default)(onCreated)) {
            onCreated(_this11.pluginMap[name]);
          }
        });
      }
    }
  }, {
    key: 'setControlBar',
    value: function setControlBar(opts) {
      var _this12 = this;

      if (this.pluginMap.ControlBar) {
        // do nothing
      } else {
        var onCreated = opts.onCreated,
            restOpts = _objectWithoutProperties(opts, ['onCreated']);

        var initOpts = _extends({}, defaultOpts.ControlBar, restOpts);
        this.map.plugin(['AMap.ControlBar'], function () {
          _this12.pluginMap.ControlBar = new window.AMap.ControlBar(initOpts);
          _this12.map.addControl(_this12.pluginMap.ControlBar);
          if ((0, _isFun2.default)(onCreated)) {
            onCreated(_this12.pluginMap.ControlBar);
          }
        });
      }
    }

    // 用户可以通过 created 事件获取 map 实例

  }, {
    key: 'exposeMapInstance',
    value: function exposeMapInstance() {
      if ('events' in this.props) {
        var events = this.props.events || {};
        if ((0, _isFun2.default)(events.created)) {
          events.created(this.map);
          delete events.created;
        }
        return events;
      }
      return false;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this13 = this;

      return React.createElement(
        'div',
        { style: wrapperStyle },
        React.createElement(
          'div',
          { ref: function ref(div) {
              _this13.mapWrapper = div;
            }, style: containerStyle },
          this.state.mapLoaded ? null : this.props.loading || null
        ),
        React.createElement(
          'div',
          null,
          this.state.mapLoaded ? this.renderChildren() : null
        )
      );
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.map.destroy();
    }
  }]);

  return Map;
}(Component);

exports.default = Map;
